pallets/loans/src/types/cashflow.rs

```diff
@@ -0,0 +1,429 @@
+pub enum InterestPayments {
+	/// All interest is expected to be paid at the maturity date
+	None,
+
+	/// Interest is expected to be paid monthly
+	/// The associated value correspond to the paydown day in the month,
+	/// from 1-31.
+	/// The day will be adjusted to the month.
+	///
+	/// NOTE: Only day 1 is supported by now
+	Monthly(u8),
+}
+
+/// Specify the paydown schedules of the loan
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum PayDownSchedule {
+	/// No restrictions on how the paydown should be done.
+	None,
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct CashflowPayment<Balance> {
+	pub when: Seconds,
+	pub principal: Balance,
+	pub interest: Balance,
+}
+
+/// Specify the repayment schedule of the loan
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct RepaymentSchedule {
+	/// Expected repayments date for remaining debt
+	pub maturity: Maturity,
+
+	/// Period at which interest is paid
+	pub interest_payments: InterestPayments,
+
+	/// How much of the initially borrowed amount is paid back during interest
+	/// payments
+	pub pay_down_schedule: PayDownSchedule,
+}
+
+impl RepaymentSchedule {
+	pub fn is_valid(&self, now: Seconds) -> Result<bool, DispatchError> {
+		match self.interest_payments {
+			InterestPayments::None => (),
+			InterestPayments::Monthly(_) => {
+				let start = date::from_seconds(now)?;
+				let end = date::from_seconds(self.maturity.date())?;
+
+				// We want to avoid creating a loan with a cashflow consuming a lot of computing
+				// time Maximum 40 years, which means a cashflow list of 40 * 12 elements
+				if end.year() - start.year() > 40 {
+					return Ok(false);
+				}
+			}
+		}
+
+		Ok(self.maturity.is_valid(now))
+	}
+
+	pub fn generate_cashflows<Balance, Rate>(
+		&self,
+		origination_date: Seconds,
+		principal: Balance,
+		interest_rate: &InterestRate<Rate>,
+	) -> Result<Vec<CashflowPayment<Balance>>, DispatchError>
+	where
+		Balance: FixedPointOperand + EnsureAdd + EnsureDiv,
+		Rate: FixedPointNumber + EnsureDiv,
+	{
+		let start_date = date::from_seconds(origination_date)?;
+		let end_date = date::from_seconds(self.maturity.date())?;
+
+		let (timeflow, periods_per_year) = match &self.interest_payments {
+			InterestPayments::None => (vec![], 1),
+			InterestPayments::Monthly(reference_day) => (
+				date::monthly_intervals(start_date, end_date, (*reference_day).into())?,
+				12,
+			),
+		};
+
+		let total_weight = timeflow
+			.iter()
+			.map(|(_, weight)| weight)
+			.try_fold(0, |a, b| a.ensure_add(*b))?;
+
+		let interest_per_period = interest_rate
+			.per_year()
+			.ensure_div(Rate::saturating_from_integer(periods_per_year))?;
+
+		timeflow
+			.into_iter()
+			.map(|(date, weight)| {
+				let proportion = FixedU128::ensure_from_rational(weight, total_weight)?;
+				let principal = proportion.ensure_mul_int(principal)?;
```

Comment:
This should always be based on the whole `principal` amount. 

Example: 
* `principal = 100`
* 12 periods, `5%` `interest_per_period` for the sake of simplicity

Without a paydown in the `principal` amount you would expect to have an interst payment of `5` each month. 

We are having secondly compounding, so the actual interest that has to be paid will be different. As the `100` are compounding every second, and then the interest to be payed is actually above `5` and will increase each month. Slightly of course only. 

My mind model is: 
if the compounding interval is the same as the repayment interval, then there is no compounding at all. If the compounding interval is smaller, then there is compounding. If the repayment interval is smaller, then there is no compounding and the interest payments each month are getting smaller.

Reply:
Then, in order to compute a cashflow upfront, we need to ensure the repayment interval is always <= than the compounding. If not, we can not compute an exact cashflow upfront, right?

Reply:
Yes, but right now, it is anyways not exact, given the secondly think. So I think this is an estimate on the given state.